_posts/2020-09-01-weather-prediction/weather_prediction_deep_learning.Rmd

```diff
@@ -0,0 +1,695 @@
+            list(pad_width, pad_width),
+            list(0L, 0L),
+            list(0L, 0L)
+          ))
+      }
+    }
+  })
+}
+
+periodic_conv_2d <- function(filters,
+                             kernel_size,
+                             name = NULL) {
+  
+  keras_model_custom(name = name, function(self) {
+    self$padding <- periodic_padding_2d(pad_width = (kernel_size - 1) / 2)
+    self$conv <-
+      layer_conv_2d(filters = filters,
+                    kernel_size = kernel_size,
+                    padding = 'valid')
+    
+    function (x, mask = NULL) {
+      x %>% self$padding() %>% self$conv()
+    }
+  })
+}
+```
+
+For our purposes of establishing a deep learning baseline that is fast to train, CNN architecture and parameter defaults are
+chosen to be simple and moderate, respectively:[^7]
+
+[^7]: These are the same filter and kernel sizes as employed in [Rasp et al.\'s simple CNN example on
+    github](https://github.com/pangeo-data/WeatherBench/blob/master/notebooks/3-cnn-example.ipynb).
+
+```{r}
+periodic_cnn <- function(filters = c(64, 64, 64, 64, 2),
+                         kernel_size = c(5, 5, 5, 5, 5),
+                         dropout = rep(0.2, 5),
+                         name = NULL) {
+  
+  keras_model_custom(name = name, function(self) {
+    
+    self$conv1 <-
+      periodic_conv_2d(filters = filters[1], kernel_size = kernel_size[1])
+    self$act1 <- layer_activation_leaky_relu()
+    self$drop1 <- layer_dropout(rate = dropout[1])
+    self$conv2 <-
+      periodic_conv_2d(filters = filters[2], kernel_size = kernel_size[2])
+    self$act2 <- layer_activation_leaky_relu()
+    self$drop2 <- layer_dropout(rate =dropout[2])
+    self$conv3 <-
+      periodic_conv_2d(filters = filters[3], kernel_size = kernel_size[3])
+    self$act3 <- layer_activation_leaky_relu()
+    self$drop3 <- layer_dropout(rate = dropout[3])
+    self$conv4 <-
+      periodic_conv_2d(filters = filters[4], kernel_size = kernel_size[4])
+    self$act4 <- layer_activation_leaky_relu()
+    self$drop4 <- layer_dropout(rate = dropout[4])
+    self$conv5 <-
+      periodic_conv_2d(filters = filters[5], kernel_size = kernel_size[5])
+    
+    function (x, mask = NULL) {
+      x %>%
+        self$conv1() %>%
+        self$act1() %>%
+        self$drop1() %>%
+        self$conv2() %>%
+        self$act2() %>%
+        self$drop2() %>%
+        self$conv3() %>%
+        self$act3() %>%
+        self$drop3() %>%
+        self$conv4() %>%
+        self$act4() %>%
+        self$drop4() %>%
+        self$conv5()
+    }
+  })
+}
+
+model <- periodic_cnn()
+```
+
+## Training
+
+In that same spirit of "default-ness", we train with MSE loss and Adam optimizer.
+
+```{r}
+loss <- tf$keras$losses$MeanSquaredError(reduction = tf$keras$losses$Reduction$SUM)
+optimizer <- optimizer_adam()
+
+train_loss <- tf$keras$metrics$Mean(name='train_loss')
+
+valid_loss <- tf$keras$metrics$Mean(name='test_loss')
+
+train_step <- function(train_batch) {
```

Comment:
it doesn't seem to me that we are using any custom training logic. Why not just using `keras %>% fit`?

Reply:
Historically, this never worked (using custom models with `fit` in TF 2.0+...) 

I just tried again and get the error
```
RuntimeError: Evaluation error: unused argument (training = TRUE). 
```

Honestly, I also think that one can't show this method often enough as it seems like it never got used much outside some small group ... just like custom models themselves ... (although who knows, depending on whether/how the new TF 2.3 pass-in-your-custom-logic will work from R, perhaps that kind of flexibility could be achieved even easier at some point...)